File: 0494-target-sum/0494-target-sum.java
class Solution {
    int dfs(int[] nums, int target, int idx, int sum, int cnt) {
        if (idx >= nums.length) {
            if (sum == target)
                return cnt+1;
            return cnt;
        }
        cnt = dfs(nums, target, idx+1, sum+nums[idx], cnt);
        cnt = dfs(nums, target, idx+1, sum-nums[idx], cnt);
        return cnt;
    }
    
    public int findTargetSumWays(int[] nums, int target) {
        int idx = 0, sum = 0, cnt = 0;
        cnt = dfs(nums, target, idx, sum, cnt);
        return cnt;
    }
}

File: 0494-target-sum/0494-target-sum.py
class Solution:
    def dfs(self, idx, target, nums, dp):
        if idx == 0:
            if target+nums[idx] == 0 and target-nums[idx] == 0:
                return 2
            if target+nums[idx] == 0 or target-nums[idx] == 0:
                return 1
            return 0
        if (idx, target) in dp:
            return dp[(idx, target)]
        plus = self.dfs(idx-1, target+nums[idx], nums, dp)
        minus = self.dfs(idx-1, target-nums[idx], nums, dp)
        dp[(idx, target)] = plus + minus
        
        return dp[(idx, target)]
    
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        dp = {}
        return self.dfs(len(nums)-1, target, nums, dp)
        